base/libgit2/callbacks.jl

```diff
@@ -29,20 +22,45 @@ function mirror_callback(remote::Ptr{Ptr{Void}}, repo_ptr::Ptr{Void},
     return Cint(0)
 end
 
+"""Credentials callback function
+
+Function provides different credential acquisition functionality w.r.t. a connection protocol.
+If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayload` object.
+
+For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields: `user` & `pass` they are used to create authentication credentials.
+In addition, if payload has field `used` it can be set to `true` to indicate that payload was used and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
+"""
 function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
                               username_ptr::Cstring,
-                              allowed_types::Cuint, payload::Ptr{Void})
+                              allowed_types::Cuint, payload_ptr::Ptr{Void})
     err = 1
     url = bytestring(url_ptr)
 
     if isset(allowed_types, Cuint(Consts.CREDTYPE_USERPASS_PLAINTEXT))
-        # use keyboard-interactive prompt
-        username = prompt("Username for '$url'")
-        pass     = prompt("Password for '$url'", password=true)
+        username = password = ""
+        if payload_ptr != C_NULL
+            payload = unsafe_pointer_to_objref(payload_ptr)
+            if isa(payload, AbstractPayload)
+                username = isdefined(payload, :user) ? getfield(payload, :user) : ""
```

Comment:
I feel that checking for specific field names for a given type should be wrapped in a generic function: `user(::Payload)`, `password(::Payload)`, `isused(::Payload)`, etc.